### ISSUE TITLE
endbr64 removal

### ISSUE THREAD

**Issue by rui314** [llvm/llvm-project#74400]
When `-fcf-protection=branch` is given, the compiler conservatively emits an `endbr64` landing pad instruction at the beginning of each globally-visible function. This is because, even if the address of a function was not taken within the current compilation unit, it might be taken in another translation unit. That means the compiler has no choice other than to conservatively emit `endbr64` to allow indirect branches.

That conservative assumption resulted in having lots of unnecessary gadgets in the linked code.

mold has recently gained the `-z rewrite-endbr` option to rewrite unnecessary `endbr64` instructions with NOPs. It was doable because the linker can do a simple whole-program analysis to determine whether an address of a function is taken or not; a function's address is considered not taken if and only if the function is referred only by `R_X86_64_PLT32`.

Would you like to implement this same feature in lld? In particular, OpenBSD guys are interested in it to tighten security of their operating system.

**Comment by MaskRay** [comment]
This is an interesting option and is like a non-LTO alternative to `-mibt-seal`.
I have been personally contemplating whether these gadget reduction mechanisms genuinely make exploitation significantly more difficult and have remained cautious.

I have some notes at https://maskray.me/blog/2022-12-18-control-flow-integrity#dont-conservatively-mark-non-local-linkage-functions

> Considering the large code model and potential inclusion of range extension thunks, a function might still be indirectly jumped to even if the preceding condition evaluates to false.
> As a precaution, we'll need to identify and label such call sites with the NOTRACK prefix.

As a side note, the current mold code may rewrite endbr for an exported symbol (e.g. `--export-dynamic`).

**Comment by rui314** [comment]
The other thing you can do today without adding any new command line flag is to emit endbr64 only to canonical PLTs. Since there's no way to get an address of a non-canonical PLT entry, they'll never be called indirectly via a function pointer and therefore don't need an endbr64 at the beginning of their code sequences.

**Comment by MaskRay** [comment]
> The other thing you can do today without adding any new command line flag is to emit endbr64 only to canonical PLTs. Since there's no way to get an address of a non-canonical PLT entry, they'll never be called indirectly via a function pointer and therefore don't need an endbr64 at the beginning of their code sequences.

Yes. I changed AArch64 BTI to use this optimization: https://reviews.llvm.org/D110217 . A non-preemptible ifunc also needs a landing instruction. Range extension thunks (https://reviews.llvm.org/D148704) and `PLT32` data relocation also need a landing instruction. At this point I am actually unsure whether x86 needs a similar "optimization".

**Comment by smithp35** [comment]
On AArch64 there is an outstanding ABI issue https://github.com/ARM-software/abi-aa/issues/196 which I need to write up in more detail. This is about range-extension thunks and BTI as there is a difference in code-generation between GCC (which can elide the BTI (endbr equivalent) when it thinks a function can't be address taken, but clang never elides the BTI.

There is a way to make range-extension thunks work, essentially splitting them into two parts, with the first part doing an indirect jump, and second a direct jump. GNU ld now uses these when necessary (requires disassembly of the call site to do optimally). Would like to see that in LLD as it would permit more BTIs to be removed.

We have thought of attempting a similar linker option to remove unnecessary BTIs but so far haven't done enough work to suggest that it is worth the extra implementation complexity in the linker, and the risk of getting some (possibly assembly generated) cases wrong and have some programs randomly crash.

**Comment by brad0** [comment]
> When `-fcf-protection=branch` is given, the compiler conservatively emits an `endbr64` landing pad instruction at the beginning of each globally-visible function. This is because, even if the address of a function was not taken within the current compilation unit, it might be taken in another translation unit. That means the compiler has no choice other than to conservatively emit `endbr64` to allow indirect branches.
> 
> That conservative assumption resulted in having lots of unnecessary gadgets in the linked code.
> 
> mold has recently gained the `-z rewrite-endbr` option to rewrite unnecessary `endbr64` instructions with NOPs. It was doable because the linker can do a simple whole-program analysis to determine whether an address of a function is taken or not; a function's address is considered not taken if and only if the function is referred only by `R_X86_64_PLT32`.
> 
> Would you like to implement this same feature in lld? In particular, OpenBSD guys are interested in it to tighten security of their operating system.

OpenBSD is interested in this and it would be nice to have upstream. We already have enough local patching in other areas.

**Comment by kettenis** [comment]
I spent a bit of time trying the implement this for lld. My first attempt follows the approach taken in mold where I add an "address taken" bit to each symbol and set it base on the relocations that reference the symbol.  Unfortunately I ran into some corner cases where things didn't work:

1. The generated code that runs the global constructors is referenced from `.init_array` using a relocation that references the section with the offset to the symbol in the added: 
```
Relocation section '.rela.init_array' at offset 0x13e8 contains 1 entries:
    Offset             Info             Type               Symbol's Value  Symbol's Name + Addend
0000000000000000  0000000200000101 R_AARCH64_ABS64        0000000000000000 .text + b0
``` 
So the corresponding symbol doesn't get its "address taken" bit set, `endbr64` gets replaced by `nop` , but since the function is called indirectly the code crashes.

2. If there are symbol aliases (two or more symbols that point at the same bit of code) and only one of the aliases has its address taken, the code ends up replacing the `endbr64` with `nop` through the other aliases.  Resulting in a crash. This happens for example in C++ destructors because clang generates aliases for those (for ABI compatibility?). But I suspect it will also happen in OpenBSD's libc where hidden aliases are created for internal references to public functions.

I looked into using the information from `llvm_addrsig`. That solves the first issue, but makes the second issue worse since C++ destructors are not considered to be address significant, even though a pointer to global destructors gets passed to `__cxa_atexit()`.

Maybe another fix for the first issue would be to "resolve" references to sections by looking if they match the location of a symbol in that section. I didn't try to implement this yet.

The second issue is harder. I suppose the code could check for aliases and mark those as "address taken" as well if one of the matching symbols has the bit set. Is there an efficient way to do this with the current data structures?

**Comment by smithp35** [comment]
A very quick drive by comment.

I vaguely remember a similar problem occurring with thunks. I think the solution in that case was to store a symbol as a <section, offset> pair https://github.com/llvm/llvm-project/blob/main/lld/ELF/Relocations.h#L174

**Comment by rui314** [comment]
For the first issue, I think we need to find the real, symbol at a given offset if the symbol referred by a relocation is a section symbol.

For the second issue, it's not too hard to find symbol aliases, no? I'd sort all symbols by their addresses, scan them from the beginning to the end, and then propagate "address taken" if a symbol at the same address is address taken.

**Comment by rui314** [comment]
@kettenis We fixed the section symbol issue in https://github.com/rui314/mold/commit/ed7eec5969078f7420a9329c792ad3f8942f3fc0. You may want to take a look at it. With that, mold can self-host with CET enabled.
